juju/juju
# Change endpoint bindings to use space ids instead of names

## Description of change

Part of the ongoing network work.  Endpoint Bindings will now be saved as Space IDs rather than Names to facilitate future changes.  Included are upgrade and migration steps.

Impact of this change:
- All EndpointBindings will have a value, if not set by the user, it will be the DefaultSpace.
- The default EndpointBindings will always be included in results from app.EndpointBindings().
- state.AllSpaces() now returns the default space.
- Output of some juju commands will now reference space ids instead of space names.  This output will be reviewed and potentially updated in the near future.

## QA steps

Run the following scenarios with MAAS, AWS, and non space aware clouds.
1. deploy a mysql unit with space bindings, then use in a CMR config.
2. deploy a unit with space bindings then migrate the unit
3. deploy a unit with space bindings then upgrade juju 
3. add a machine to a specific space.
3. deploy units with no bindings.
3. deploy units with memory constraints
3. deploy a unit with space bindings and export a bundle, verify bundle.
3. changing bindings during charm upgrade.
